src/abstracttypes.jl

```diff
@@ -0,0 +1,31 @@
+"""
+    AbstractHypergraph{T} <: AbstractMatrix{Union{T, Nothing}}
+
+An abstract hypergraph type storing information about vertices and hyperedges.
+"""
+# TODO: should this be just `abstract type AbstractHypergraph{T} end`?
+# How important is subtyping AbstractMatrix?
+abstract type AbstractHypergraph{T} <: AbstractMatrix{T} end
+
+"""
+    AbstractSimpleHypergraph{T} <: AbstractHypergraph{T}
+
+An abstract undirected hypergraph type storing information about vertices and hyperedges.
+"""
+abstract type AbstractSimpleHypergraph{T} <: AbstractHypergraph{T} end
+
+
+# fundamental traits
+# Inspired by HyperGraphs.jl
+@traitdef IsDirected{X <: AbstractHypergraph}
+@traitdef HasMeta{X <: AbstractHypergraph}
+
+# functions and default behaviour: direction
+@traitimpl IsDirected{T} <- isdirected(T)
+isdirected(::Type{T}) where {T} = false
+isdirected(X::T) where {T} = isdirected(T)
+
+# functions and default behaviour: vertex/hyperedge metadata
+@traitimpl HasMeta{T} <- hasmeta(T)
```

Comment:
We have metadata both for vertices and edges. These are denoted in `Hypergraph` constructor by parametric types `V` and `E` respectively. 
Hence it should be `HasVertexMeta` and `HasHyperedgeMeta` (I know hypergraph libraries that eg allow only meta for hyperedges)
Currently if there is no meta-data the respective parametric type `V` or `E` is `Nothing`. I assume this can added to the `hasvertexmeta` and `hashyperedgemeta` implementations for the `Hypergraph` type.

Reply:
Sure, can make this change.